File: strategies/tradepro_strategies/cli/trade_odds.py
"""Trade odds — "if I place a limit at X and target Y, what are the chances?"

Owner: "we should be able to place an order at 920 and close at 950 and let it
run ... calculate the probability."

WHAT THIS IS. A base rate measured on this symbol's OWN history, not a
forecast. It walks every historical bar, pretends the same order was resting
there, and counts what actually happened. If MU reached a +3.3% target before
a -8% stop in 61% of past attempts, that is the number — no model, no
distribution assumption, no opinion.

WHY IT IS TWO QUESTIONS, NOT ONE. This is the part the phrasing hides. With MU
at 966.78, a limit at 920 is a bet that price falls 4.8% FIRST. Most of the
time it simply will not, and you are flat. So the honest answer is a chain:

    P(filled)                  does price reach the limit at all, within N days
    P(target | filled)         and THEN does it reach the target before the stop
    P(both)                    the two multiplied — what actually happens to you

Quoting only the second number would flatter the trade badly.

CONSERVATIVE ON AMBIGUITY. Daily bars cannot say whether the high or the low
came first. When one session touches BOTH the target and the stop, this counts
it as a STOP. That biases every number DOWNWARD, which is the correct
direction for something a person will risk money on.

WHAT IT DELIBERATELY DOES NOT USE.

  * Live prices. Daily bars up to the last settled session. Nothing here is
    a live quote and it must never be described as one.
  * Volume. Momentum v3 gated on entry volume and was REJECTED — the apparent
    edge exists only in the second half of the record and inverts before 2020
    (MOMENTUM_GATES_V3.md, e50cd2a). Using it here would contradict our own
    evidence.
  * Support and resistance levels. Tested over 76,260 touch events, both
    edges negative against random placebo lines. What replaces it is the
    EXCURSION table: how far this symbol actually travels before reversing,
    measured rather than drawn.

REGIME WARNING. The all-history number and the recent number are both shown
and they can differ a lot. MU trading at 88 in 2025 and 966 now is not the
same instrument's behaviour. Trust the recent row more, and the sample size
above all.
"""
from __future__ import annotations

import argparse
import logging
import statistics as st

from .momentum_candidates import _load, poison_check, _tradeable

log = logging.getLogger("tradepro.trade_odds")


def barrier_scan(c, h, l, dates, *, limit_pct, target_pct, stop_pct,
                 fill_window, trade_window, start_idx=0):
    """Replay a resting limit order from every historical bar.

    limit_pct/target_pct/stop_pct are all RELATIVE TO THE PRICE AT THE MOMENT
    THE ORDER IS PLACED, so the same order shape can be measured across a
    history spanning an 11x price change.
    """
    attempts = filled = won = lost = timed = 0
    bars_to_fill: list[int] = []
    bars_to_win: list[int] = []
    outcomes: list[float] = []

    n = len(c)
    for i in range(start_idx, n - 1):
        ref = c[i]
        if ref <= 0:
            continue
        attempts += 1
        limit = ref * (1 + limit_pct)
        fill_j = None
        for j in range(i + 1, min(n, i + fill_window + 1)):
            # A limit BELOW the market fills when price trades down to it;
            # a limit ABOVE fills when price trades up to it.
            if (limit_pct <= 0 and l[j] <= limit) or (limit_pct > 0 and h[j] >= limit):
                fill_j = j
                break
        if fill_j is None:
            continue
        filled += 1
        bars_to_fill.append(fill_j - i)

        tgt = limit * (1 + target_pct)
        stp = limit * (1 + stop_pct)
        done = False
        for j in range(fill_j, min(n, fill_j + trade_window + 1)):
            hit_t = h[j] >= tgt
            hit_s = l[j] <= stp
            if hit_t and hit_s:
                # Same session touched both. Daily bars cannot order them, so
                # assume the bad one — never the flattering one.
                lost += 1; outcomes.append(stop_pct * 100); done = True; break
            if hit_t:
                won += 1; bars_to_win.append(j - fill_j)
                outcomes.append(target_pct * 100); done = True; break
            if hit_s:
                lost += 1; outcomes.append(stop_pct * 100); done = True; break
        if not done:
            timed += 1
            end = min(n - 1, fill_j + trade_window)
            outcomes.append(100 * (c[end] / limit - 1))

    return {
        "attempts": attempts, "filled": filled, "won": won, "lost": lost,
        "timed_out": timed,
        "p_fill": (filled / attempts) if attempts else None,
        "p_target_given_fill": (won / filled) if filled else None,
        "p_stop_given_fill": (lost / filled) if filled else None,
        "p_both": (won / attempts) if attempts else None,
        "median_bars_to_fill": (st.median(bars_to_fill) if bars_to_fill else None),
        "median_bars_to_target": (st.median(bars_to_win) if bars_to_win else None),
        "mean_outcome_pct": (round(st.mean(outcomes), 2) if outcomes else None),
    }


def excursion_table(c, h, l, *, window, start_idx=0):
    """How far does this symbol actually TRAVEL before reversing?

    The empirical answer to the question support/resistance lines are usually
    asked. From each bar, the maximum favourable move within `window`
    sessions. If the median is +4% and you are targeting +12%, the target is
    fighting the symbol's own behaviour — and that is measured, not drawn.
    """
    ups: list[float] = []
    n = len(c)
    for i in range(start_idx, n - 1):
        if c[i] <= 0:
            continue
        end = min(n, i + window + 1)
        ups.append(100 * (max(h[i + 1:end]) / c[i] - 1))
    if not ups:
        return None
    ups.sort()
    q = lambda p: ups[int(p * (len(ups) - 1))]  # noqa: E731
    return {"n": len(ups), "p25": round(q(.25), 1), "median": round(q(.50), 1),
            "p75": round(q(.75), 1), "p90": round(q(.90), 1)}


def sweep_targets(c, h, l, dates, *, limit_pct, stop_pct, fill_window,
                  trade_window, start_idx=0, targets=None):
    """Same order, swept across a range of TARGETS.

    Because the target is the variable actually worth solving. MU shows why:
    a +3.3% target hits 83% of the time, which reads as a great trade until
    you notice the -8% stop means one loss erases 2.4 wins — while the symbol
    typically travels +18% in the same window. A high hit rate and a good
    trade are different things, and only expectancy tells them apart.

    Expectancy per FILLED order is the column to read.
    """
    targets = targets or [1, 2, 3, 5, 8, 12, 18, 25]
    rows = []
    for t in targets:
        r = barrier_scan(c, h, l, dates, limit_pct=limit_pct, target_pct=t / 100,
                         stop_pct=stop_pct, fill_window=fill_window,
                         trade_window=trade_window, start_idx=start_idx)
        rows.append({
            "target_pct": t,
            "p_target_given_fill": r["p_target_given_fill"],
            "expectancy_pct": r["mean_outcome_pct"],
            "reward_risk": round(t / abs(stop_pct * 100), 2) if stop_pct else None,
            "median_bars_to_target": r["median_bars_to_target"],
            "filled": r["filled"],
        })
    return rows


def compute(symbol, *, entry, target, stop_pct, fill_window, trade_window, ref=None):
    df = _load(symbol)
    if df is None:
        return {"error": f"no stored daily bars for {symbol}"}
    c = df["close"].tolist(); h = df["high"].tolist(); l = df["low"].tolist()
    dates = [str(x)[:10] for x in df.index]
    ok, ratio = poison_check(c)
    if not ok:
        return {"error": f"{symbol} is quarantined — historical max is {ratio}x the recent "
                         f"median, consistent with a wrong-venue series. No odds computed."}
    last = c[-1]
    ref = ref if ref is not None else last
    limit_pct = entry / ref - 1
    target_pct = target / entry - 1
    if target_pct <= 0:
        return {"error": "target must be above the entry"}

    # Recent = last ~2 years of sessions, shown separately because regime
    # matters more than sample size here (momentum v3 proved that the hard way).
    recent_start = max(0, len(c) - 504)
    out = {
        "symbol": symbol, "as_of_bar": dates[-1], "last_close": round(last, 2),
        "reference_price": round(ref, 2), "entry": entry, "target": target,
        "limit_move_pct": round(100 * limit_pct, 2),
        "target_move_pct": round(100 * target_pct, 2),
        "stop_pct": round(100 * stop_pct, 2),
        "stop_price": round(entry * (1 + stop_pct), 2),
        "fill_window": fill_window, "trade_window": trade_window,
        "all_history": barrier_scan(c, h, l, dates, limit_pct=limit_pct,
                                    target_pct=target_pct, stop_pct=stop_pct,
                                    fill_window=fill_window, trade_window=trade_window),
        "recent_2y": barrier_scan(c, h, l, dates, limit_pct=limit_pct,
                                  target_pct=target_pct, stop_pct=stop_pct,
                                  fill_window=fill_window, trade_window=trade_window,
                                  start_idx=recent_start),
        "excursion_all": excursion_table(c, h, l, window=trade_window),
        "excursion_recent": excursion_table(c, h, l, window=trade_window,
                                            start_idx=recent_start),
        # BOTH eras, always. A sweep on the last two years alone would have
        # told an MU trader to target +25%, because MU rose 11x over exactly
        # that window. Momentum v3 was rejected for believing a second-half
        # regime was an edge; the same mistake is available here and is far
        # more expensive, because a person acts on this one directly.
        "sweep_all": sweep_targets(c, h, l, dates, limit_pct=limit_pct,
                                   stop_pct=stop_pct, fill_window=fill_window,
                                   trade_window=trade_window),
        "sweep_recent": sweep_targets(c, h, l, dates, limit_pct=limit_pct,
                                      stop_pct=stop_pct, fill_window=fill_window,
                                      trade_window=trade_window, start_idx=recent_start),
        "history_from": dates[0], "bars": len(c),
        # Said out loud rather than left to be inferred from a small number in
        # a corner. SKHY has two partitions of history; odds computed on it
        # would look identical in shape to MU's 4,183 bars and mean nothing.
        "thin_history": (None if len(c) >= 500 else
                         f"only {len(c)} stored sessions from {dates[0]} — too little history "
                         f"for these odds to mean much. Treat them as indicative at best."),
        "not_live": ("Computed from stored daily bars up to the last settled session. "
                     "These are not live prices and this is a base rate, not a forecast."),
    }
    return out


def _pct(x):
    return "—" if x is None else f"{100 * x:.0f}%"


def main() -> int:
    ap = argparse.ArgumentParser(description="Historical odds for a limit-entry trade")
    ap.add_argument("symbol")
    ap.add_argument("--entry", type=float, required=True, help="limit price you would rest")
    ap.add_argument("--target", type=float, required=True)
    ap.add_argument("--stop-pct", type=float, default=-8.0, help="stop, %% from entry (default -8)")
    ap.add_argument("--fill-window", type=int, default=10, help="sessions to wait for a fill")
    ap.add_argument("--trade-window", type=int, default=20, help="sessions to hold once filled")
    a = ap.parse_args()
    logging.basicConfig(level=logging.WARNING)

    r = compute(a.symbol.upper(), entry=a.entry, target=a.target,
                stop_pct=a.stop_pct / 100, fill_window=a.fill_window,
                trade_window=a.trade_window)
    if "error" in r:
        print(f"✗ {r['error']}")
        return 1
    print(f"{r['symbol']} — last settled close {r['last_close']} ({r['as_of_bar']})")
    print(f"  rest a limit at {r['entry']} ({r['limit_move_pct']:+.1f}%), "
          f"target {r['target']} ({r['target_move_pct']:+.1f}% from entry), "
          f"stop {r['stop_price']} ({r['stop_pct']:+.1f}%)")
    print(f"  wait up to {r['fill_window']} sessions for a fill, then hold up to {r['trade_window']}\n")
    print(f"{'':<12}{'attempts':>9}{'P(fill)':>9}{'P(tgt|fill)':>13}{'P(BOTH)':>9}{'avg out':>9}")
    for k, label in (("all_history", "all history"), ("recent_2y", "last 2yr")):
        s = r[k]
        print(f"{label:<12}{s['attempts']:>9}{_pct(s['p_fill']):>9}"
              f"{_pct(s['p_target_given_fill']):>13}{_pct(s['p_both']):>9}"
              f"{(str(s['mean_outcome_pct']) + '%') if s['mean_outcome_pct'] is not None else '—':>9}")
    for k, label in (("excursion_all", "all history"), ("excursion_recent", "last 2yr")):
        e = r[k]
        if e:
            print(f"\nhow far it travels up within {r['trade_window']} sessions ({label}): "
                  f"median {e['median']}% · 75th {e['p75']}% · 90th {e['p90']}%")
    print(f"\n{'':>8}{'':>7}{'--- all history ---':>26}{'--- last 2 years ---':>26}")
    print(f"{'target':>8}{'R:R':>7}{'P(hit|fill)':>13}{'expectancy':>13}"
          f"{'P(hit|fill)':>13}{'expectancy':>13}")
    best_a = max((x["expectancy_pct"] or -99) for x in r["sweep_all"])
    best_r = max((x["expectancy_pct"] or -99) for x in r["sweep_recent"])
    for ra, rr in zip(r["sweep_all"], r["sweep_recent"]):
        mark = ""
        if ra["expectancy_pct"] == best_a and rr["expectancy_pct"] == best_r:
            mark = "  ← best in BOTH"
        elif ra["expectancy_pct"] == best_a:
            mark = "  ← best all-history"
        elif rr["expectancy_pct"] == best_r:
            mark = "  ← best recent only"
        print(f"{('+' + str(ra['target_pct']) + '%'):>8}{ra['reward_risk']:>7}"
              f"{_pct(ra['p_target_given_fill']):>13}{(str(ra['expectancy_pct']) + '%'):>13}"
              f"{_pct(rr['p_target_given_fill']):>13}{(str(rr['expectancy_pct']) + '%'):>13}{mark}")
    print("  A target that is best ONLY in the recent column is a bet on the regime "
          "continuing,\n  not a property of the symbol.")
    if r.get("thin_history"):
        print(f"\n⚠ {r['thin_history']}")
    print(f"\n{r['not_live']}")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: frontend/src/lib/tradeOdds.ts
/**
 * Trade odds — barrier maths, ported from
 * `strategies/tradepro_strategies/cli/trade_odds.py`.
 *
 * TWO IMPLEMENTATIONS OF THE SAME MATHS, deliberately. The Python CLI is the
 * reference and is what studies are run through; this runs in the browser so
 * the screen can be interactive without a Python service behind it (the API is
 * C#). They are validated against each other on a known case — see
 * `tradeOdds.test.ts`. If you change one, change both and re-check.
 *
 * The rules that matter, and why:
 *
 *  - A limit BELOW the market is a bet price falls to you FIRST. So the answer
 *    is a chain — P(filled), then P(target | filled) — never just the second.
 *  - Daily bars cannot say whether the high or the low came first. A session
 *    touching BOTH barriers is counted as a STOP. Every number is therefore
 *    biased downward, which is the right direction for money.
 *  - Everything is measured in PERCENT from the price at order time, so a
 *    history spanning an 11x price change stays comparable.
 */
export type Bar = { ts: string; open: number; high: number; low: number; close: number };

export type Scan = {
  attempts: number; filled: number; won: number; lost: number; timedOut: number;
  pFill: number | null; pTargetGivenFill: number | null; pBoth: number | null;
  medianBarsToTarget: number | null; meanOutcomePct: number | null;
};

const median = (xs: number[]) =>
  xs.length ? [...xs].sort((a, b) => a - b)[Math.floor(xs.length / 2)] : null;

export function barrierScan(
  bars: Bar[],
  opts: { limitPct: number; targetPct: number; stopPct: number;
          fillWindow: number; tradeWindow: number; startIdx?: number },
): Scan {
  const { limitPct, targetPct, stopPct, fillWindow, tradeWindow } = opts;
  const start = opts.startIdx ?? 0;
  let attempts = 0, filled = 0, won = 0, lost = 0, timedOut = 0;
  const barsToWin: number[] = [];
  const outcomes: number[] = [];
  const n = bars.length;

  for (let i = start; i < n - 1; i++) {
    const ref = bars[i].close;
    if (!(ref > 0)) continue;
    attempts++;
    const limit = ref * (1 + limitPct);
    let fillJ = -1;
    for (let j = i + 1; j < Math.min(n, i + fillWindow + 1); j++) {
      if ((limitPct <= 0 && bars[j].low <= limit) || (limitPct > 0 && bars[j].high >= limit)) {
        fillJ = j; break;
      }
    }
    if (fillJ < 0) continue;
    filled++;

    const tgt = limit * (1 + targetPct);
    const stp = limit * (1 + stopPct);
    let done = false;
    for (let j = fillJ; j < Math.min(n, fillJ + tradeWindow + 1); j++) {
      const hitT = bars[j].high >= tgt;
      const hitS = bars[j].low <= stp;
      if (hitT && hitS) { lost++; outcomes.push(stopPct * 100); done = true; break; }
      if (hitT) { won++; barsToWin.push(j - fillJ); outcomes.push(targetPct * 100); done = true; break; }
      if (hitS) { lost++; outcomes.push(stopPct * 100); done = true; break; }
    }
    if (!done) {
      timedOut++;
      const end = Math.min(n - 1, fillJ + tradeWindow);
      outcomes.push(100 * (bars[end].close / limit - 1));
    }
  }
  const mean = outcomes.length ? outcomes.reduce((a, b) => a + b, 0) / outcomes.length : null;
  return {
    attempts, filled, won, lost, timedOut,
    pFill: attempts ? filled / attempts : null,
    pTargetGivenFill: filled ? won / filled : null,
    pBoth: attempts ? won / attempts : null,
    medianBarsToTarget: median(barsToWin),
    meanOutcomePct: mean === null ? null : Math.round(mean * 100) / 100,
  };
}

/** How far the symbol actually TRAVELS up within the window — the measured
 *  answer to the question support/resistance lines are usually asked. */
export function excursion(bars: Bar[], window: number, startIdx = 0) {
  const ups: number[] = [];
  for (let i = startIdx; i < bars.length - 1; i++) {
    if (!(bars[i].close > 0)) continue;
    let hi = -Infinity;
    for (let j = i + 1; j < Math.min(bars.length, i + window + 1); j++) hi = Math.max(hi, bars[j].high);
    if (hi > -Infinity) ups.push(100 * (hi / bars[i].close - 1));
  }
  if (!ups.length) return null;
  ups.sort((a, b) => a - b);
  const q = (p: number) => Math.round(ups[Math.floor(p * (ups.length - 1))] * 10) / 10;
  return { n: ups.length, p25: q(0.25), median: q(0.5), p75: q(0.75), p90: q(0.9) };
}

export const DEFAULT_TARGETS = [1, 2, 3, 5, 8, 12, 18, 25];

export function sweepTargets(
  bars: Bar[],
  opts: { limitPct: number; stopPct: number; fillWindow: number; tradeWindow: number;
          startIdx?: number; targets?: number[] },
) {
  return (opts.targets ?? DEFAULT_TARGETS).map((t) => {
    const r = barrierScan(bars, { ...opts, targetPct: t / 100 });
    return {
      targetPct: t,
      pTargetGivenFill: r.pTargetGivenFill,
      expectancyPct: r.meanOutcomePct,
      rewardRisk: opts.stopPct ? Math.round((t / Math.abs(opts.stopPct * 100)) * 100) / 100 : null,
      medianBarsToTarget: r.medianBarsToTarget,
      filled: r.filled,
    };
  });
}


/**
 * DIP-FROM-OPEN scan — the owner's own strategy shape.
 *
 * Different from `barrierScan` in one way that matters: the limit is placed
 * relative to each session's OPEN, not to one reference price. That is how the
 * order would actually be worked — you decide "1% below wherever it opens"
 * every morning, not a fixed price.
 *
 * SAME-SESSION AMBIGUITY. If the session's low reaches the limit AND its high
 * clears the target, the trade only worked if the low came first, and daily
 * bars cannot say. `pessimistic` (the default, and what the study graded on)
 * assumes it did not, and carries the position. The owner spotted this
 * unprompted: "the high might have hit early then we placed the order and then
 * it never hit that high in that day."
 *
 * EXPECTANCY IS THE ANSWER, NOT WIN RATE. Backtested at a 0.5% target against
 * an 8% stop this wins 66% of the time and loses 0.41% per trade, because it
 * needs 94% to break even. The UI must show both or it teaches the wrong
 * lesson. See INTRADAY_DIP_GATES_V1.md — REJECTED.
 */
export function dipScan(
  bars: Bar[],
  opts: { dipPct: number; targetPct: number; stopPct: number; carryDays: number;
          pessimistic?: boolean; startIdx?: number },
) {
  const { dipPct, targetPct, stopPct, carryDays } = opts;
  const pess = opts.pessimistic ?? true;
  const MAXDAY = 0.35;
  let sessions = 0, fills = 0, sameDay = 0;
  const res: number[] = [], days: number[] = [];
  const n = bars.length;
  let i = Math.max(1, opts.startIdx ?? 1);
  while (i < n) {
    const o = bars[i].open, pc = bars[i - 1].close;
    if (!(o > 0) || !(pc > 0) || Math.abs(o / pc - 1) > MAXDAY) { i++; continue; }
    sessions++;
    const limit = o * (1 - dipPct / 100);
    if (bars[i].low > limit) { i++; continue; }   // never traded down to us
    fills++;
    const T = limit * (1 + targetPct / 100), S = limit * (1 + stopPct / 100);
    let out: [number, number] | null = null;
    for (let j = i; j < Math.min(n, i + carryDays + 1); j++) {
      if (j > i && (!(bars[j - 1].close > 0) || Math.abs(bars[j].close / bars[j - 1].close - 1) > MAXDAY)) break;
      let hitT = bars[j].high >= T;
      const hitS = bars[j].low <= S;
      if (j === i && pess) hitT = false;
      if (hitS) { out = [stopPct, j - i]; break; }
      if (hitT) { out = [targetPct, j - i]; break; }
    }
    if (!out) {
      const j = Math.min(n - 1, i + carryDays);
      out = [100 * (bars[j].close / limit - 1), j - i];
    }
    res.push(out[0]); days.push(Math.max(1, out[1]));
    if (out[1] === 0) sameDay++;
    i += Math.max(1, out[1]) + 1;
  }
  if (!res.length) return null;
  const totalDays = days.reduce((a, b) => a + b, 0);
  const wins = res.filter((x) => x > 0).length;
  return {
    sessions, fills, trades: res.length,
    fillRate: sessions ? fills / sessions : 0,
    winRate: (100 * wins) / res.length,
    expPerTrade: res.reduce((a, b) => a + b, 0) / res.length,
    // Capital-time weighting: total return over total days COMMITTED. Not
    // mean(return/days), which overweights short lucky trades — that error
    // reported a 11.6x result that was really 5.08x.
    expPerDayHeld: totalDays ? res.reduce((a, b) => a + b, 0) / totalDays : 0,
    meanHold: totalDays / res.length,
    sameDayPct: (100 * sameDay) / res.length,
  };
}

/** Being long open→close, the thing any of this has to beat. */
export function benchmarkPerDay(bars: Bar[]) {
  const r: number[] = [];
  for (let i = 1; i < bars.length; i++) {
    const o = bars[i].open, pc = bars[i - 1].close;
    if (o > 0 && pc > 0 && Math.abs(bars[i].close / pc - 1) <= 0.35) r.push(100 * (bars[i].close / o - 1));
  }
  return r.length ? r.reduce((a, b) => a + b, 0) / r.length : null;
}

/**
 * THE STRATEGY RULES, evaluated against one symbol on demand.
 *
 * Owner: "what if I want to see the probability on a particular symbol — will
 * I just put that symbol in odds and it will run all the strategy with latest
 * data for it."
 *
 * It did not, and that was a fair expectation to have. Odds answered "if I
 * place THIS order, how often did it work" — it never asked what our own
 * strategies think of the name.
 *
 * These are ports of `signals/mean_reversion.py` and the momentum entry in
 * `cli/momentum_candidates.py`. They are checked against the Python by
 * `strategies/scripts/check-trade-odds-parity.sh` — if you change one, change
 * both and re-run it. A drifting copy of an entry rule is how a screen starts
 * disagreeing with the backtest that justified it.
 *
 * Each returns not just fires/does-not-fire but HOW FAR from firing, because
 * "no" is the answer on almost every symbol on almost every day, and a bare
 * "no" tells you nothing about whether to look again tomorrow.
 */
const mean = (xs: number[]) => xs.reduce((a, b) => a + b, 0) / xs.length;
const smaAt = (c: number[], i: number, n: number) => mean(c.slice(i - n + 1, i + 1));

export type RuleCheck = {
  fires: boolean;
  headline: string;
  clauses: Array<{ label: string; value: string; ok: boolean }>;
  plan?: { entry: number; target?: number; stop: number; targetPct?: number };
};

/** Swing — 2.5σ below the 20-day mean, while above the 200-day average. */
export function checkSwing(bars: Bar[]): RuleCheck | null {
  const c = bars.map((b) => b.close);
  const i = c.length - 1;
  if (i < 210) return null;
  const w = c.slice(i - 19, i + 1);
  const m20 = mean(w);
  const sd = Math.sqrt(mean(w.map((x) => (x - m20) ** 2)));   // population sd, as Python
  const s200 = smaAt(c, i, 200);
  const band = m20 - 2.5 * sd;
  const belowBand = c[i] < band;
  const aboveTrend = c[i] > s200;
  const sigmasBelow = sd > 0 ? (m20 - c[i]) / sd : 0;
  const fires = belowBand && aboveTrend;
  return {
    fires,
    // A negative "sigmasBelow" means the price is ABOVE its mean, and reading
    // "-2.2σ below" is worse than useless — TSLA sits 2.2σ ABOVE and would
    // have read as almost qualifying. Say which side of the mean it is on.
    headline: fires
      ? `FIRES — ${sigmasBelow.toFixed(1)}σ below the 20-day mean, above the 200-day average`
      : sigmasBelow < 0
        ? `no — trading ${Math.abs(sigmasBelow).toFixed(1)}σ ABOVE its 20-day mean; this rule buys dips`
        : `no — only ${sigmasBelow.toFixed(1)}σ below the mean, needs 2.5σ${aboveTrend ? "" : ". It is also BELOW its 200-day average, which blocks the rule outright"}`,
    clauses: [
      { label: "At least 2.5σ below the 20-day mean", ok: belowBand,
        value: `${sigmasBelow >= 0 ? "" : "+"}${Math.abs(sigmasBelow).toFixed(2)}σ ${sigmasBelow >= 0 ? "below" : "ABOVE"} · close ${c[i].toFixed(2)} vs band ${band.toFixed(2)}` },
      { label: "Above the 200-day average (the trend floor)", ok: aboveTrend,
        value: `${(100 * (c[i] / s200 - 1)).toFixed(1)}% · close ${c[i].toFixed(2)} vs ${s200.toFixed(2)}` },
    ],
    plan: { entry: c[i], target: m20, stop: c[i] * 0.92,
            targetPct: 100 * (m20 / c[i] - 1) },
  };
}

/** Momentum — a pullback TO the 10-day average inside a confirmed uptrend. */
export function checkMomentum(bars: Bar[]): RuleCheck | null {
  const c = bars.map((b) => b.close);
  const i = c.length - 1;
  if (i < 210) return null;
  const s200 = smaAt(c, i, 200), s50 = smaAt(c, i, 50);
  const s20 = smaAt(c, i, 20), s10 = smaAt(c, i, 10);
  const prev10 = smaAt(c, i - 1, 10);
  const cl = [
    { label: "Above the 200-day average", ok: c[i] > s200,
      value: `${(100 * (c[i] / s200 - 1)).toFixed(1)}%` },
    { label: "20-day above the 50-day (uptrend)", ok: s20 > s50,
      value: `${(100 * (s20 / s50 - 1)).toFixed(1)}%` },
    { label: "Still above the 20-day average", ok: c[i] > s20,
      value: `${(100 * (c[i] / s20 - 1)).toFixed(1)}%` },
    { label: "Pulled back TO the 10-day average (trigger)", ok: c[i] <= s10 * 1.005,
      value: `${(100 * (c[i] / s10 - 1)).toFixed(1)}% (fires at ≤ +0.5%)` },
    { label: "Was above it yesterday — a pullback, not a breakdown", ok: c[i - 1] > prev10,
      value: `${(100 * (c[i - 1] / prev10 - 1)).toFixed(1)}%` },
  ];
  const fires = cl.every((x) => x.ok);
  const failed = cl.filter((x) => !x.ok);
  return {
    fires,
    headline: fires ? "FIRES — pullback to the 10-day average in an uptrend"
      : `no — ${failed.length} of 5 conditions unmet (${failed[0].label.toLowerCase()})`,
    clauses: cl,
    plan: { entry: c[i], stop: c[i] * 0.92 },
  };
}
